### PR TITLE
[ONNX] Introduce ONNXScopeName

### LINKED ISSUE #75100: Preserve original scope names in exported ONNX graph
### 🚀 The feature, motivation and pitch

When a PyTorch model is exported, the graph nodes take uninformative names like `f"{op_type}_{op_cnt}"`, and so do tensors (besides the input and output tensors of the overall graph).
In some contexts, it would be extremely helpful if each node in the exported ONNX graph would be named after the scope name of the PyTorch module from which it came from. Additionally, intermediate tensors should also ideally be named after the node that created them.

Here's a very subjective list of why it matters:

### Visualisation, Debugging

netron_view

Let's say one goal of exporting to ONNX is to visualise a model (without loss of generality, let's think Netron) to reason about its architecture. For anything but the most trivial NN, this is simply not possible if nodes are named `Conv_{N}` and tensors are `{M}`. 

As soon as the node name becomes `unet.encoder.stage1.conv0` and its output tensor `unet.encoder.stage1.conv0.output0`, I become able to understand what node am I looking at, where is a skip connection going, and this sort of things.

### ONNX graph manipulation

For some people, exporting a model to ONNX is not the end of the story. There may be some graph- or weight- level manipulations that must be applied on sub-graphs defined by the name scopes of their nodes or their intermediate tensors: just to give an idea, a silly example here may be "apply batchnorm fusion only on the decoder of the UNet".

### TensorFlow does it 😛 

Let's be clear: I am by no means a TF fan, but credit is due here. On that side, the feature I'm pitching is the default in [`tf2onnx`](https://github.com/onnx/tensorflow-onnx/tree/main/tf2onnx) and is a design decision that makes a lot of sense.

### Notes

I fully understand that name scopes may lose some of their meaning when running graph-level optimisations during the export if sets of nodes are replaced by new ones and there is no 1:1 relationship. Even in this case, though, it may be possible to assign a "common ancestor name"... or fall back to the current implementation if nothing can be done.

### 1. Fork PyTorch and diverge

Since at least some of the code of the ONNX exporting is in C++, there is no easy monkey patching solution, the library must be recompiled.

### 2. Hack around with a heavy hand

[Someone](https://github.com/pytorch/pytorch/issues/33463#issuecomment-606399944) suggested that it is possible to monkey patch the `_slow_forward` method of `torch.nn.Module` to give meaningful names to the _tensors_ flying around during graph tracing (note: not the nodes, but from output tensor names to node names the step is trivial).
I tried it and it seems to be a bit of a hit and miss, I would need to have a way deeper understanding of the guts of JIT to ensure that 100% of the output tensors actually have a useful name.

Alternatively, or in conjunction, I am also trying to monkey patch `torch.onnx.utils._export()` to use the `inlined_graph`, dump it, and rename the ONNX nodes as a post-processing.

### Additional context

The scope name would be readily available from the `scopeName()` of each `torch._C.Node` of the TorchScript's `inlined_graph`, but the export instead uses the `graph` where such information has been stripped off (for example, [here](https://github.com/pytorch/pytorch/blob/00607e75e5f3157839719efb5b26ca92e91190a3/torch/onnx/utils.py#L430) is the first place where the choice of using the `graph` is made and becomes consequential when exporting a `ScriptModule`).

To make use of this metadata down the line, also the serialisation logic of `torch._C.Graph._export_onnx()` must be adapted. [Here](https://github.com/pytorch/pytorch/blob/00607e75e5f3157839719efb5b26ca92e91190a3/torch/csrc/jit/serialization/export.cpp#L830) is where the nodes of the exported ONNX proto take their names on the fly.

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82040
* #82039
* __->__ #82038

Update `_setup_trace_module_map` to always record module/layer info
in `Scope` attribute for nodes.
Extend `Scope` name to not only record module typename, but also
module object variable name. Both names are formatted and stored
as `name` attribute in `Scope`.
Introduce `ONNXScopeName` class to manage the formatting and parsing.
Updated local function export code adjusting to this update.

### CODE REVIEW

File: torch/onnx/utils.py [hunk 1]
```diff
@@ -1251,12 +1251,10 @@ def unconvertible_ops(
...
-def _setup_trace_module_map(model, export_modules_as_functions):
-    def __setup_trace_module_map():
-        trace_module_map = {_m: torch.typename(type(_m)) for _m in model.modules()}
-        torch.jit._trace._trace_module_map = trace_module_map
-        return trace_module_map
-
+def _setup_trace_module_map(
+    model: Union[torch.nn.Module, torch.jit.ScriptModule],
+    export_modules_as_functions: Union[bool, Collection[Type[torch.nn.Module]]],
+):
```
Comment (justinchuby) [inline torch/onnx/utils.py:1257]:
```suggestion
) -> bool:
```
does it return bool?

File: torch/onnx/utils.py [hunk 2]
```diff
@@ -1280,9 +1278,15 @@ def _track_module_attributes_forward_hook(module, input, output):
...
+    trace_module_map = {
+        _m: torch._C._jit_onnx_create_full_scope_name(torch.typename(type(_m)), _n)
+        for _n, _m in model.named_modules()
+    }
+    torch.jit._trace._trace_module_map = trace_module_map
     if isinstance(export_modules_as_functions, bool) and export_modules_as_functions:
-        trace_module_map = __setup_trace_module_map()
-        export_modules_as_functions = {v for k, v in trace_module_map.items()}
+        export_modules_as_functions = {
+            torch.typename(type(_m)) for _m, _ in trace_module_map.items()
```
Comment (justinchuby) [inline torch/onnx/utils.py:1288]:
```suggestion
            torch.typename(type(named_module)) for named_module in trace_module_map
```

maybe this for simplicity?

File: torch/onnx/utils.py [hunk 3]
```diff
@@ -1280,9 +1278,15 @@ def _track_module_attributes_forward_hook(module, input, output):
...
+        }
     elif (
         isinstance(export_modules_as_functions, set)
         and len(export_modules_as_functions) > 0
```
Comment (justinchuby) [inline torch/onnx/utils.py:1292]:
```suggestion
        and export_modules_as_functions
```
nit: unrelated to this pr but

File: torch/onnx/utils.py (hunk 1 above)
Comment (AllenTiTaiWang) [reply torch/onnx/utils.py]: I think it's not. From the if and elif branch, it seems like could be dict.

File: torch/onnx/utils.py [hunk 4]
```diff
@@ -1298,11 +1302,10 @@ def _find_typename(v):
                     "Got `%s`." % (type(v).__name__)
                 )
 
-        trace_module_map = __setup_trace_module_map()
         module_typenames = {_find_typename(v) for v in export_modules_as_functions}
         export_modules_as_functions = module_typenames
     else:
-        export_modules_as_functions = None
+        export_modules_as_functions = False
```
Comment (AllenTiTaiWang) [inline torch/onnx/utils.py]: I wonder if there is a reason `export_modules_as_functions` should be a bool if it's not a dict? I see it's later consumed by `_jit_pass_onnx_function_extraction`

Comment (BowenBao) [reply torch/onnx/utils.py]: I think it was a mistake. Now it's renamed, and assigned as empty set.

**Review by titaiwangms — APPROVED** [review]
:beers:

File: torch/onnx/utils.py (hunk 1 above)
Comment (BowenBao) [reply torch/onnx/utils.py]: updated to `Set[str]`

File: torch/csrc/jit/passes/onnx/naming.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+namespace ONNXScopeName {
+
+using NameFunc = std::string (*)(torch::jit::ScopePtr scope);
+
+const std::string name_separator = "::";
+
+namespace {
+
+std::string nameFromRoot(
```
Comment (justinchuby) [inline torch/csrc/jit/passes/onnx/naming.cpp:16]: Should functions follow PascalCase naming?

**Review by justinchuby — APPROVED** [review]

File: torch/csrc/jit/passes/onnx/naming.cpp (hunk 5 above)
Comment (justinchuby) [reply torch/csrc/jit/passes/onnx/naming.cpp]: I realized PyTorch uses camel case for functions everywhere else. Sorry for making the previous suggestion

**Comment by BowenBao** [comment]
cc @abock for merge approval

**Review by abock — APPROVED** [review]

**Comment by BowenBao** [comment]
@malfet could you help merge this one? I think it's touching `build_variables.bzl`

**Review by malfet — APPROVED** [review]
